### PR TITLE
Elula Peer Review

### PR DESCRIPTION
**Pull request by tharradine**
This is where I'll put my feedback for the code review.

### CODE REVIEW

File: src/addrmatcher/matcher.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+            )
+
+        if similarity_threshold < 0:
+            raise ValueError("Similarity threshold has to be larger than 0")
+
+        if nlargest < 1:
+            raise ValueError("The number of returned records must be at least 1")
+
+        # initiate the result
+        addresses = pd.DataFrame()
```
Comment (tharradine) [inline src/addrmatcher/matcher.py:425]: Looks like this variable is not used

File: src/addrmatcher/matcher.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+                        / 100.0
+                    )
+                else:
+                    address_parquet["RATIO"] = address_parquet["FULL_ADDRESS"].apply(
+                        lambda x: jaro_winkler_similarity(
+                            re.sub(r"[\W_]+", "", address.upper()),
+                            re.sub(r"[\W_]+", "", x.upper()),
+                        )
+                        / 100.0
+                    )
```
Comment (tharradine) [inline src/addrmatcher/matcher.py:520]: This looks very similar to the code above. You could make this drier by extracting the code into a separate function or method.

File: src/addrmatcher/region.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
+"""
+Data structure for the regional unit
+"""
+
+
+class Region:
```
Comment (tharradine) [inline src/addrmatcher/region.py:6]: I would say this class is a pretty good candidate for a Python dataclass, will make this file much smaller and simpler :)

Comment (tharradine) [reply src/addrmatcher/region.py:6]: Also, this is a mutable object with simple getters and setters for every attribute. This is fairly redundant in Python, we tend to just use public attributes. I know in other OOP languages (like Java) the convention is to use private attributes and add getters and setters, but in Python this is generally considered overkill.

The only real reason for using properties in Python is to:
- Make an attribute read-only (the property has no setter)
- Give a mutable attribute a custom setter, i.e. something other than just `self._attr = attr`

File: src/addrmatcher/__main__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from .resource import main
```
Comment (tharradine) [inline src/addrmatcher/__main__.py:1]: I think you meant to import `download`?

File: src/addrmatcher/matcher.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+            raise ValueError(
+                "The latitude input should be within "
+                + self._hierarchy.coordinate_boundary[0]
+                + "and"
+                + self._hierarchy.coordinate_boundary[1]
+                + " and longitude input must be within  "
+                + self._hierarchy.coordinate_boundary[2]
+                + "and"
+                + self._hierarchy.coordinate_boundary[3]
+            )
```
Comment (tharradine) [inline src/addrmatcher/matcher.py:681]: Looks like you're trying to concatenate strings and floats here. f-strings would prevent this issue, and also probably make it a bit more readable.

File: src/addrmatcher/matcher.py [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+        df = pq.read_table(
+            self._filename,
+            filesystem=local,
+            filters=[
+                ("LATITUDE", ">=", lat - distance),
+                ("LATITUDE", "<=", lat + distance),
+                ("LONGITUDE", ">=", lon - distance),
+                ("LONGITUDE", "<=", lon + distance),
+            ],
+        ).to_pandas()
```
Comment (tharradine) [inline src/addrmatcher/matcher.py:611]: Specifying a filesystem here shouldn't be necessary, pyarrow should default to a local filesystem

**Review by tharradine — COMMENTED** [review]
Hi guys,

Firstly, great work on this package, it's been put together quite nicely and I can see you've put a lot of thought and effort into it.

Please see inline comments for specific parts of the code where I can see room for improvement. Some of those comments I've made on a single instance of an issue which appears multiple times, so try to look for places where those issues are repeated :)

Here is also some general feedback which doesn't necessarily apply to specific parts of the code.

### Docstrings
Your docstrings are pretty good, but some of them need more clarification around specifics, or edge cases. They tend to describe what the method *should* do, but don't explain what happens when something unexpected happens, e.g. if there's no match for an address, or if there's a missing file, the user needs to know how to handle that situation.

When you write a docstring, always ask yourself, "if I were a user, could I read this docstring and confidently know exactly how to use this method / class / function"? It should include information on not just what the thing should do, but also things like what happens when things go wrong, what exceptions are raised, what specific data-types should each argument be (not just 'list', more specific like 'list of string', 'list of Region' etc.), and what is returned under *any* circumstance, not just the successful circumstance.

### Testing
This package could be vastly improved with some good unit tests. The tests shouldn't just do the high-level address matching, but they should test each function (or "unit") individually, e.g. cleaning addresses, traversing the geography hierarchy, etc.

pytest is a really good tool for doing this but it's not the only framework out there.

### Downloading data
The data download workflow for a user is not super intuitive, and you've written a bunch of code in the `resource.py` module which could probably be avoided. I think the integration of data into the package could be more seamless if each country had its data vendored in separate packages on PyPI. Then you could enable support for Australia with something like `pip install addrmatcher[AUS]`, which would install the `addrmatcher-data-AUS` package as well, which contains all the data files. I suppose for some countries, the region data could be too big to fit on PyPI, in which case I suppose your method works better, but I still think the data should be downloaded to an anonymous, shared location somewhere, not just any directory the user can specify and remember.

### Documentation
It would be good to have some documentation on the actual steps the address matcher goes through to find a match. I understand string-matching is involved, and some address cleaning can be done, but I'm not 100% sure on the full process, and this is something documentation could help explain.

**Comment by HninPwint** [comment]
Hi @tharradine,

Thank you for the detail review of our code.

The data download workflow was the least favourable option for us. Our first choice was to integrate data into the package. When we faced the size limit issue from PyPI, we opted to data download option temporarily.

We have contacted PyPi-support team and made a request to increase the size limit for our package. [https://github.com/pypa/pypi-support](url). If it is approved, we can drop the data download workflow.

Cheers!!!

File: src/addrmatcher/matcher.py [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+            A matched address
+        """
+
+        # initiate the dataframe
+        matched_df = pd.DataFrame()
+
+        # search State and Postcode and extract - example QLD 410
+        match = re.search(
+            r"\s((?:NSW|VIC|QLD|TAS|WA|SA|NT|ACT))\s([0-9]{4})$", no_number_address
+        )
```
[the remark this answers is not in the document]
Comment (isampe) [reply src/addrmatcher/matcher.py:190]: This line is part of the cleaning process, which is currently applied to Australian addresses only (e.g., revise the invalid locality name). However, the package can match non-Australian addresses, but the cleaning parameter must be set to False.

image

For the time being, we found no non-Australian address, so it's quite difficult for us to make the cleaning process more generic, as it is highly dependant on the region's structure.

However, we can discuss how to make the cleaning process more generic.

**Pull request closed without merging**